source/agora/test/EnrollmentManager.d

```diff
@@ -221,3 +221,188 @@ unittest
     // Check if a new pre-image has been revealed from the restarted node
     assert(preimage_2.isValid(org_preimage, validator_cycle));
 }
+
+/// Situation: One misbehaving node sends an Enrollment for an already-enrolled
+///            validator.
+/// Expectation: The nomination is rejected.
+unittest
+{
+    import agora.common.Config;
+    import agora.common.Task;
+    import agora.common.crypto.Key;
+    import agora.consensus.protocol.Nominator;
+    import agora.consensus.data.ConsensusData;
+    import agora.node.Ledger;
+    import agora.node.Validator;
+    import agora.network.NetworkManager;
+
+    import geod24.Registry;
+
+    static class BadNominator : Nominator
+    {
+        private Ledger ledger;
+        private bool is_nominating;
+
+        /// Ctor
+        public this (NetworkManager network, KeyPair key_pair,
+            Ledger ledger, TaskManager taskman)
+        {
+            this.ledger = ledger;
+            super(network, key_pair, ledger, taskman);
+        }
+
+        ///
+        public override bool isNominating () @safe @nogc nothrow
+        {
+            return this.is_nominating;
+        }
+
+        ///
+        public override void tryNominate () @safe
+        {
+            // Checks done in `Nominator.tryNominate`
+            if (this.is_nominating)
+                return;
+
+            this.is_nominating = true;
+            scope (exit) this.is_nominating = false;
+
+            ConsensusData data;
+            this.ledger.prepareNominatingSet(data);
+            if (data.tx_set.length == 0)
+                return;  // not ready yet
+
+            // Add an enrollment that is already an validator
+            data.enrolls ~= GenesisBlock.header.enrollments[0];
+            // data.enrolls ~= Enrollment.init;
+
+            // note: we are not passing the previous tx set as we don't really
+            // need it at this point (might later be necessary for chain upgrades)
```

Comment:
Likewise, remove the comment, it doesn't matter here (it's copy-pasted code, that's what matters).